### PR TITLE
Make Windows Widgets Compatible w/ Other Platforms

### PR DESCRIPTION
**Pull request by time-killer-games**
This is a complete re-write of Windows Widgets that adds UTF-8 support to most of the functions as well as adds all the functions that were always a part of the other system that were missing in Win32 to be a part of Windows Widgets as well. You may test this with the following demo EXE and GM81 source files: [Win32.zip](https://github.com/enigma-dev/enigma-dev/files/3296744/Win32.zip)

As a side note, I changed the get_directory() function to have "Select Directory" as the titlebar caption and the "OK" button label to say "Select" to match what YoYo Games did with the exact same dialog in the IDE:
Untitled
Robert complained last time I changed that dialog to do this and I'm just explaining my reasoning for this.

### CODE REVIEW

**Comment by time-killer-games** [comment]
@JoshDreamland I want you to review this and run the demo EXE with Wine because Robert isn't as familiar with how the other Widgets Systems are made as well as you are. Since the whole point of this PR is to make Windows Widgets compatible and identical to our other Widget Systems I figured you'd have better judgement on this being both the founder/lead developer of the project and a Linux user. Robert doesn't have access to Linux so he's no help with comparing this to the Linux Widget Systems.

**Comment by time-killer-games** [comment]
@JoshDreamland travis got fucked right in the pussy

**Comment by time-killer-games** [comment]
@RobertBColton could you please not cancel my travis k thx

**Comment by time-killer-games** [comment]
@RobertBColton @JoshDreamland Alright guys, it's now passing. Let me know what you think when you get the chance. It is ready for review. If you can't review Robert, please let Josh know that so he can do it instead.

**Comment by time-killer-games** [comment]
@RobertBColton at the time of this post, I've been made aware you too upset with something unrelated to me so dont worry about addressing this right away and please don't feel any pressure.

There is a problem with get_open_filenames() and get_open_filenames_ext() where if you try to open files from the root of any given drive, i.e. C:\, I:\, etc. then the most amount of files that can be returned from a given selection is up to 3. If you select files found not in the root of any given drive, i.e. C:\Windows\, it will work as intended. 

I was wondering if you'd be willing to help me get to the bottom of why it is doing that and perhaps find a fix for it. Thanks!

File: ENIGMAsystem/SHELL/Universal_System/Extensions/Asynchronous/ASYNCdialog.cpp [hunk 1]
```diff
@@ -124,27 +124,27 @@ namespace enigma_user {
     return queue_async_job(fnc);
   }
 
-  int get_string_async(string message, string def, string cap) {
+  int get_string_async(string str, string def) {
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/Asynchronous/ASYNCdialog.cpp]: Why did you rename `message` to `str`? It's less descriptive.

Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/Asynchronous/ASYNCdialog.cpp]: That's what GM uses, and that is what we are using in the Mac and Linux Systems.

Comment (JoshDreamland) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/Asynchronous/ASYNCdialog.cpp]: I'd sooner change it in those systems. It's very obvious what `message` means in this context, but naming a parameter to `get_string` just `str` is incredibly helpful. It would as well not have a name.

Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/Asynchronous/ASYNCdialog.cpp]: I'll change it, but first I want to see if there are any other critiques you might have which I'll probably address all them sometime tmr.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp [hunk 2]
```diff
@@ -191,80 +273,272 @@ static INT_PTR CALLBACK ShowMessageExtProc(HWND hwndDlg, UINT uMsg, WPARAM wPara
...
 
+  return "";
 }
 
-void message_mouse_color(int col) {
+static inline string get_open_filenames_helper(string filter, string fname, string dir, string title) {
+  OPENFILENAMEW ofn;
 
-}
+  filter = string(filter).append("||");
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: This block of code is identical to the one above, with exception of the `OFN_ALLOWMULTISELECT` flag. All other differences are purely syntactical.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp [hunk 3]
```diff
@@ -191,80 +273,272 @@ static INT_PTR CALLBACK ShowMessageExtProc(HWND hwndDlg, UINT uMsg, WPARAM wPara
...
+  ofn.hwndOwner = enigma::hWnd;
+  ofn.lpstrFile = wstr_fname;
+  ofn.nMaxFile = MAX_PATH;
+  ofn.lpstrFilter = tstr_filter.c_str();
+  ofn.nFilterIndex = 0;
+  ofn.lpstrTitle = tstr_title.c_str();
+  ofn.lpstrInitialDir = tstr_dir.c_str();
+  ofn.Flags = OFN_EXPLORER | OFN_PATHMUSTEXIST | OFN_HIDEREADONLY | OFN_OVERWRITEPROMPT | OFN_NOCHANGEDIR;
+
+  if (GetSaveFileNameW(&ofn) != 0)
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: This block is identical to `get_open_filename_helper`, except it calls `GetSaveFileNameW` instead of `GetSaveFileNameW`.

**Comment by time-killer-games** [comment]
@JoshDreamland ok i fixed the repeated code issue with the file dialogs. Used another helper.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 3 above)
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Ok, I think I finally fixed it.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 2 above)
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Ok, I think I finally fixed it.

File: ENIGMAsystem/SHELL/Universal_System/Extensions/Asynchronous/ASYNCdialog.cpp (hunk 1 above)
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/Asynchronous/ASYNCdialog.cpp]: I'll make that change in a different pr directly after this one is merged, to keep the changes more organized.
